File: algorithms/greedy/kruskalAlgorithm.js
// Kruskal's Algorithm is a greedy algorithm used to find the Minimum Spanning Tree (MST) of a connected, undirected, and weighted graph.

// A Minimum Spanning Tree is a subset of the edges that connects all vertices in the graph with the minimum total edge weight, and without forming any cycles.

// Disjoint Set (Union-Find) class with path compression and union by rank
class DisjointSet {
  constructor(n) {
    // Initially, each vertex is its own parent (self-loop), forming n disjoint sets
    this.parent = Array.from({ length: n }, (_, i) => i);
    // Rank is used to keep the tree shallow (optimization)
    this.rank = new Array(n).fill(0);
  }

  // Find the root of the set that element 'u' belongs to
  find(u) {
    if (this.parent[u] !== u) {
      // Path compression: flatten the tree as we go up
      this.parent[u] = this.find(this.parent[u]);
    }
    return this.parent[u];
  }

  // Union two sets if they are not already connected
  union(u, v) {
    const rootU = this.find(u);
    const rootV = this.find(v);

    // If both vertices belong to the same set, adding this edge would create a cycle
    if (rootU === rootV) return false;

    // Union by rank: attach the smaller tree to the root of the larger one
    if (this.rank[rootU] < this.rank[rootV]) {
      this.parent[rootU] = rootV;
    } else if (this.rank[rootU] > this.rank[rootV]) {
      this.parent[rootV] = rootU;
    } else {
      this.parent[rootV] = rootU;
      this.rank[rootU]++; // Increase rank if both trees have same height
    }

    return true; // Union was successful
  }
}

// Main function to apply Kruskal's Algorithm
const kruskalAlgorithm = (verticesCount, edges) => {
  // Step 1: Sort edges by their weights in non-decreasing order
  edges.sort((a, b) => a[2] - b[2]);

  const ds = new DisjointSet(verticesCount); // Create disjoint sets for all vertices
  const mst = []; // To store edges of the Minimum Spanning Tree

  // Step 2: Loop through sorted edges and pick valid ones
  for (const [u, v, weight] of edges) {
    // If including this edge doesn't form a cycle
    if (ds.union(u, v)) {
      mst.push([u, v, weight]); // Add it to MST
    }

    // Optimization: If MST has (V - 1) edges, it's complete
    if (mst.length === verticesCount - 1) break;
  }

  return mst; // Return the final MST
};

// Number of vertices in graph
const vertices = 7;

// Vertex1 , Vertex2, Weight
const edges = [
  [0, 1, 7],
  [0, 3, 5],
  [1, 2, 8],
  [1, 3, 9],
  [1, 4, 7],
  [2, 4, 5],
  [3, 4, 15],
  [3, 5, 6],
  [4, 5, 8],
  [4, 6, 9],
  [5, 6, 11],
];

const mst = kruskalAlgorithm(vertices, edges);

console.log("Minimum Spanning Tree:");
mst.forEach(([u, v, w]) => console.log(`Edge ${u}-${v} with weight ${w}`));

File: data-structures/Graph/KruskalAlgorithm.java
import java.util.ArrayList;
import java.util.Collections;

// Represents an edge in the graph
class Edge implements Comparable<Edge> {
    int source;
    int destination;
    int weight;

    public Edge(int source, int destination, int weight) {
        this.source = source;
        this.destination = destination;
        this.weight = weight;
    }

    // Sort edges in ascending order of weight
    @Override
    public int compareTo(Edge e) {
        return this.weight - e.weight;
    }
}

public class KruskalAlgorithm {

    // Stores all the edges of the graph
    static ArrayList<Edge> edges = new ArrayList<>();

    // Parent array for the Disjoint Set (Union-Find)
    static int[] parent = new int[7];

    // Creates the weighted undirected graph
    static void createGraph() {

        // Edge 3 <-> 5
        edges.add(new Edge(3, 5, 1));
        edges.add(new Edge(5, 3, 1));

        // Edge 3 <-> 4
        edges.add(new Edge(3, 4, 2));
        edges.add(new Edge(4, 3, 2));

        // Edge 4 <-> 5
        edges.add(new Edge(4, 5, 5));
        edges.add(new Edge(5, 4, 5));

        // Edge 5 <-> 6
        edges.add(new Edge(5, 6, 8));
        edges.add(new Edge(6, 5, 8));

        // Edge 0 <-> 1
        edges.add(new Edge(0, 1, 10));
        edges.add(new Edge(1, 0, 10));

        // Edge 2 <-> 4
        edges.add(new Edge(2, 4, 10));
        edges.add(new Edge(4, 2, 10));

        // Edge 1 <-> 3
        edges.add(new Edge(1, 3, 12));
        edges.add(new Edge(3, 1, 12));

        // Edge 0 <-> 2
        edges.add(new Edge(0, 2, 15));
        edges.add(new Edge(2, 0, 15));
    }

    // Initially, every vertex is its own parent
    static void initialize() {
        for (int i = 0; i < parent.length; i++) {
            parent[i] = i;
        }
    }

    // Finds the representative (parent) of a set
    static int find(int x) {

        // If the vertex is its own parent, it is the representative of the set
        if (parent[x] == x) {
            return x;
        }

        // Continue searching until the representative is found
        return find(parent[x]);
    }

    // Merges two different sets
    static void union(int a, int b) {

        // Find representatives of both sets
        int parentA = find(a);
        int parentB = find(b);

        // Merge the two sets
        parent[parentB] = parentA;
    }

    static void kruskal() {

        // Initialize the Disjoint Set
        initialize();

        // Sort all edges by increasing weight
        Collections.sort(edges);

        // Stores the total cost of the MST
        int mstCost = 0;

        // Traverse all edges in sorted order
        for (Edge e : edges) {

            // Find the representative of both vertices
            int parentSource = find(e.source);
            int parentDestination = find(e.destination);

            // If both vertices belong to different sets,
            // adding this edge will not create a cycle
            if (parentSource != parentDestination) {

                // Merge the two sets
                union(parentSource, parentDestination);

                // Add the edge weight to the MST cost
                mstCost += e.weight;

                // Print the selected edge
                System.out.println(
                        e.source + " -> " +
                        e.destination +
                        " Weight = " +
                        e.weight);
            }
        }

        // Print the total cost of the MST
        System.out.println("Total MST Cost = " + mstCost);
    }

    public static void main(String[] args) {

        // Create the graph
        createGraph();

        // Find the Minimum Spanning Tree
        kruskal();
    }
}